File: examples/js/run_wasm.mjs
/**
 * Node.js에서 WASI WebAssembly 모듈 실행 예제
 * 
 * 사용법:
 *   node run_wasm.mjs                              # say_hello() 실행
 *   node run_wasm.mjs <wasm_path>                  # 지정 파일로 say_hello() 실행
 *   node run_wasm.mjs <wasm_path> --list-root     # list_root_directory() 호출
 * 
 * 요구사항:
 *   Node.js 20+ (WASI 지원)
 * 
 * 참고: 이 예제는 reactor 모드로 빌드된 WASM 라이브러리를 사용합니다.
 *       (hello_world_wasi_lib - main() 없이 함수만 export)
 */

import { readFile } from 'node:fs/promises';
import { WASI } from 'node:wasi';
import { argv, env } from 'node:process';
import { fileURLToPath } from 'node:url';
import { dirname, resolve } from 'node:path';

// 스크립트 디렉토리 기준 경로 계산
const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);
const dummyRootPath = resolve(__dirname, '../dummy_root');

// 커맨드 라인 인자 파싱
const args = process.argv.slice(2);
const listRootFlag = args.includes('--list-root');
const wasmPath = args.find(a => !a.startsWith('--')) || './bazel-bin/main/hello_world_wasi_lib';

// WASI 인스턴스 생성 (더미 루트 디렉토리를 "/"로 매핑)
const wasi = new WASI({
  version: 'preview1',
  args: ['wasm'],
  env: env,
  preopens: listRootFlag ? { '/': dummyRootPath } : {},
});

try {
  // WASM 바이너리 로드
  const wasmBuffer = await readFile(wasmPath);
  
  // WebAssembly 모듈 컴파일 및 인스턴스화
  const wasmModule = await WebAssembly.compile(wasmBuffer);
  const instance = await WebAssembly.instantiate(wasmModule, wasi.getImportObject());
  
  // WASI 초기화 (reactor 모드 - _start 없음)
  wasi.initialize(instance);
  
  if (listRootFlag) {
    // list_root_directory 함수 직접 호출
    console.log('=== list_root_directory() 호출 ===');
    const listRootDir = instance.exports.list_root_directory;
    if (listRootDir) {
      listRootDir();
    } else {
      console.error('오류: list_root_directory 함수가 export되지 않았습니다.');
      process.exit(1);
    }
  } else {
    // say_hello 함수 호출
    const sayHello = instance.exports.say_hello;
    if (sayHello) {
      sayHello();
    } else {
      console.error('오류: say_hello 함수가 export되지 않았습니다.');
      process.exit(1);
    }
  }
  
} catch (error) {
  console.error('WASM 실행 오류:', error.message);
  process.exit(1);
}

File: examples/python/run_wasm.py
#!/usr/bin/env python3
"""
Python에서 WASI WebAssembly 모듈 실행 예제

사용법:
    python run_wasm.py
    python run_wasm.py <wasm_file_path>

요구사항:
    pip install wasmtime
"""

import sys
from pathlib import Path

try:
    from wasmtime import Engine, Store, Module, Linker, WasiConfig
except ImportError:
    print("wasmtime 패키지가 필요합니다.")
    print("설치: pip install wasmtime")
    sys.exit(1)


def run_wasm(wasm_path: str) -> int:
    """WASI WASM 모듈을 실행합니다."""
    
    # 엔진 및 스토어 생성
    engine = Engine()
    store = Store(engine)
    
    # WASI 설정
    wasi_config = WasiConfig()
    wasi_config.inherit_stdout()
    wasi_config.inherit_stderr()
    wasi_config.inherit_stdin()
    store.set_wasi(wasi_config)
    
    # 링커에 WASI 함수들 추가
    linker = Linker(engine)
    linker.define_wasi()
    
    # WASM 모듈 로드
    wasm_bytes = Path(wasm_path).read_bytes()
    module = Module(engine, wasm_bytes)
    
    # 인스턴스 생성 및 실행
    instance = linker.instantiate(store, module)
    
    # _start 함수 실행 (WASI 진입점)
    start = instance.exports(store).get("_start")
    if start is None:
        print("오류: _start 함수를 찾을 수 없습니다.")
        return 1
    
    try:
        start(store)
        return 0
    except Exception as e:
        # WASI exit(0)은 예외로 처리됨
        if "exit" in str(e).lower():
            return 0
        raise


def main():
    # WASM 파일 경로 결정
    if len(sys.argv) > 1:
        wasm_path = sys.argv[1]
    else:
        # 기본 경로
        script_dir = Path(__file__).parent
        wasm_path = script_dir / "../../bazel-bin/main/hello_world_wasi_c"
    
    wasm_path = Path(wasm_path).resolve()
    
    if not wasm_path.exists():
        print(f"오류: WASM 파일을 찾을 수 없습니다: {wasm_path}")
        print("\n먼저 빌드하세요:")
        print("  ./bazel build //main:hello_world_wasi_c --config=wasi")
        sys.exit(1)
    
    print(f"WASM 파일 로드: {wasm_path}")
    print("-" * 40)
    
    exit_code = run_wasm(str(wasm_path))
    sys.exit(exit_code)


if __name__ == "__main__":
    main()
